folding/src/standard_config.rs

```diff
@@ -0,0 +1,471 @@
+impl<G, Col, Chall, Sel, Str, I, W> FoldingEnv<G::ScalarField, I, W, Col, Chall, Sel>
+    for Env<G, Col, Chall, Sel, Str, I, W>
+where
+    G: CommitmentCurve,
+    I: Instance<G> + Index<Chall, Output = G::ScalarField> + Clone,
+    W: Witness<G> + Clone,
+    W: Index<Col, Output = Vec<G::ScalarField>> + Index<Sel, Output = Vec<G::ScalarField>>,
+    Col: FoldingColumnTrait + Eq + Hash,
+    Sel: Copy,
+    Str: Clone + Index<Col, Output = Vec<G::ScalarField>>,
+{
+    type Structure = Str;
+
+    fn new(structure: &Self::Structure, instances: [&I; 2], witnesses: [&W; 2]) -> Self {
+        // cloning for now, ideally should work with references, but that requires deeper
+        // refactorings of folding
+        let instances = instances.map(Clone::clone);
+        let witnesses = witnesses.map(Clone::clone);
+        let structure = structure.clone();
+        Self {
+            instances,
+            witnesses,
+            structure,
+            next_evals: ColumnMemoizer::new(),
+            _todo: PhantomData,
+        }
+    }
+
+    fn challenge(&self, challenge: Chall, side: Side) -> G::ScalarField {
+        let instance = match side {
+            Side::Left => &self.instances[0],
+            Side::Right => &self.instances[1],
+        };
+        // handled through Index in I
+        instance[challenge]
+    }
+
+    fn col(&self, col: Col, curr_or_next: CurrOrNext, side: Side) -> &Vec<G::ScalarField> {
+        let witness = match side {
+            Side::Left => &self.witnesses[0],
+            Side::Right => &self.witnesses[1],
+        };
+        // this should hold as long the Index implementations are consistent with the
+        // FoldingColumnTrait implementation.
+        // either search in witness for witness columns, or in the structure otherwise
+        if col.is_witness() {
+            match curr_or_next {
+                CurrOrNext::Curr => &witness[col],
+                CurrOrNext::Next => {
+                    let f = || {
+                        // simple but not the best, ideally there would be a single vector,
+                        // where you push its first element and offer either evals[0..] or
+                        // evals[1..].
+                        // that would relatively easy to implement in a custom implementation
+                        // with just a small change to this trait, but in this generic implementation
+                        // it is harder to implement.
+                        // The cost is mostly the cost of a clone
+                        let evals = &witness[col];
+                        let mut next = Vec::with_capacity(evals.len());
+                        next.extend(evals[1..].iter());
+                        next.push(evals[0]);
+                        next
+                    };
+                    self.next_evals.get_or_insert(col, f)
+                }
+            }
+        } else {
+            &self.structure[col]
```

Comment:
ok nice, that's for the selectors the `is_witness`. Can we document that in the doc of `is_witness`, please?

Reply:
I don't think it should be in `is_witness`, as this property allowing to use `is_witness` to decide between witness and structure is more a particularity of this config, and not a general requirement for any witness/columns.